### PR TITLE
Flame Graph for Profiling Data

### PR DESCRIPTION
**Pull request by MichaelMauderer**
### Pull Request Description
Add an API to create a flame graph from profiling data. Also adds a demo scene showcasing the functionality that generates some profiling data by measuring dummy function calls and rendering a flame graph for the dummy data (see video for the result).

Not that the functionality is not yet exposed user-facing in the GUI itself, but only as API and demo scene, therefore [ci no changelog needed]

https://user-images.githubusercontent.com/1428930/155118977-ecac0628-777c-48bd-9aa7-30ee6aef1976.mp4

### Important Notes
* Change from the initial design: labels are shown on the flame graph instead of as a tooltip. This is because tooltips are currently only implemented in the graph editor and would require some additional refactoring (probably taking the better part of a day). 
* re-instated the behaviour that logs are shown in the JS console if development mode is active. 

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

### CODE REVIEW

File: lib/rust/ensogl/component/flame-graph/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/component/flame-graph/src/lib.rs]: what is the reason for all these newlines in the middle of this function? I understand that you want to separate visual creation of things and adding them as children, but I think we should use fewer of such visual indicators (or none at all). What do you think?

File: lib/rust/ensogl/component/flame-graph/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/component/flame-graph/src/lib.rs]: Why is this not a separate file? As a separate file it will have better indentation and will also make this file way shorter.

File: lib/rust/ensogl/example/profiling-run-graph/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/example/profiling-run-graph/src/lib.rs]: this does not have documentation neither a name that describes what it really does.

File: lib/rust/ensogl/example/profiling-run-graph/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/example/profiling-run-graph/src/lib.rs]: Why do we need to pass `_profiler: profiler::Objective` here? Is it not designed this way that we should not do it anymore? Or will it change after the other PR gets merged?

File: lib/rust/profiler/src/flame_graph.rs
Comment (wdanilo) [inline lib/rust/profiler/src/flame_graph.rs]: Remove the name of the module from the documentation. It makes the files way harder to rename.

**Review by wdanilo — APPROVED** [review]
Amazing to see it alive! As this touches profiler too, @kazcw it would be cool if you'd review the code as well :)

File: app/ide-desktop/lib/content/src/index.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/src/index.ts:938]: Please describe what it does :)

File: lib/rust/ensogl/example/profiling-run-graph/src/lib.rs
Comment (MichaelMauderer) [reply lib/rust/ensogl/example/profiling-run-graph/src/lib.rs]: Yes, this will change after the other PR is merged. This PR will be updated then to work with the new changes.

**Comment by kazcw** [comment]
## Current (pre-#3270) `profiler` data model vs. design

The demo shows a Chrome-style-callgraph. This is not the [type of graph in the design doc](https://github.com/enso-org/design/blob/main/epics/profiling/implementation.md#run-graphs), which I've referred to there as a run-graph (out of a set of tasks, it shows which one is running).

Specifically, the difference between the two graphs is:
- In a Chrome-style-callgraph, a block represents the lifetime of a profiler, and rows indicate stack depth.
- In a run-graph, there is one row for each profiler (or profiler lifetime[^1]); blocks within a row show the intervals that that profiler is active.

[^1]: Side note: it's not clear from the image in the design doc whether a row contains the active-intervals for all lifetimes of a particular profiler, or each time that profiler is created it gets its own row. I think the former is a better default, but in the future we'd want to support distinguishing multiple lifetimes of a profiler, maybe with an expand-widget that breaks down a profiler into a different row for each lifetime, that is only used when one profiler has multiple overlapping lifetimes.

The demo reflects the currently-implemented data model: **Before #3270, `profiler` did not log enough information to generate a run-graph. However, once we have await timings, it will be possible to implement a graph like in the design doc.**

## Extend the design?

Only the run-graph can make clear the lifetimes and relationships of async tasks; however, a callgraph is better at showing non-async code (a run-graph would spread stack frames across too many rows, and not stack them nicely by depth). I think we should **revise the design to include both types of graph**.

One approach to creating a hybrid would be:
- Render a *run-graph of the async tasks*, providing a high-level view.
- Below it, render a *callgraph of non-async profilers*.
- When a block in the run-graph is clicked, zoom to that interval in the callgraph (is this easy to implement?).

## Effects on implementation

Although the data is organized differently between the two graph types, they are rendered in almost the same shapes; **the current implementation is almost capable of rendering either**. In order to support full run-graphs, it would need a container-block type to show lifetimes: a lifetime is rendered as a container-block that spans part of a row, and has leaf blocks inside it. (In the design doc image, lifetimes are light gray, with green blocks inside them.)

**Comment by MichaelMauderer** [comment]
The PR is updated to the new profiling API. @kazcw A review of the integration and changes to the profiling API would be very appreciated.

Note that the previous use of `now` failed in the IDE, and I thus changed it to us our own API wrapper instead. 


> ## Current (pre-#3270) `profiler` data model vs. design
> 
> The demo shows a Chrome-style-callgraph. This is not the [type of graph in the design doc](https://github.com/enso-org/design/blob/main/epics/profiling/implementation.md#run-graphs), which I've referred to there as a run-graph (out of a set of tasks, it shows which one is running).
> 
> Specifically, the difference between the two graphs is:
> 
>     * In a Chrome-style-callgraph, a block represents the lifetime of a profiler, and rows indicate stack depth.
> 
>     * In a run-graph, there is one row for each profiler (or profiler lifetime[1](#user-content-fn-1-1793f4dc309c9cec8d687af7f8cc7a0b)); blocks within a row show the intervals that that profiler is active.
> 
> 
> The demo reflects the currently-implemented data model: **Before #3270, `profiler` did not log enough information to generate a run-graph. However, once we have await timings, it will be possible to implement a graph like in the design doc.**
> ## Extend the design?
> 
> Only the run-graph can make clear the lifetimes and relationships of async tasks; however, a callgraph is better at showing non-async code (a run-graph would spread stack frames across too many rows, and not stack them nicely by depth). I think we should **revise the design to include both types of graph**.
> 
> One approach to creating a hybrid would be:
> 
>     * Render a _run-graph of the async tasks_, providing a high-level view.
> 
>     * Below it, render a _callgraph of non-async profilers_.
> 
>     * When a block in the run-graph is clicked, zoom to that interval in the callgraph (is this easy to implement?).
> 
> 
> ## Effects on implementation
> 
> Although the data is organized differently between the two graph types, they are rendered in almost the same shapes; **the current implementation is almost capable of rendering either**. In order to support full run-graphs, it would need a container-block type to show lifetimes: a lifetime is rendered as a container-block that spans part of a row, and has leaf blocks inside it. (In the design doc image, lifetimes are light gray, with green blocks inside them.)
> ## Footnotes
> 
>     1. Side note: it's not clear from the image in the design doc whether a row contains the active-intervals for all lifetimes of a particular profiler, or each time that profiler is created it gets its own row. I think the former is a better default, but in the future we'd want to support distinguishing multiple lifetimes of a profiler, maybe with an expand-widget that breaks down a profiler into a different row for each lifetime, that is only used when one profiler has multiple overlapping lifetimes. [leftwards_arrow_with_hook](#user-content-fnref-1-1793f4dc309c9cec8d687af7f8cc7a0b)

Sorry for the confusion, there was a design discussion with Wojciech, I have created a new PR to the design documents with the results. We wanted to see the implementation first.

You make some excellent points about improvements to the visualization, but I think we should have that discussion as part of a new PR to the design, while this PR creates a first useful literation.

File: lib/rust/profiler/flame-graph/src/lib.rs
Comment (kazcw) [inline lib/rust/profiler/flame-graph/src/lib.rs:51]: This will not produce an icicle graph, because an async task is not active during the intervals its children are.

For some async tasks **a**, **b**, **c**, **d**; where:
a.children = [b, c]
c.children = [d]

This would make a graph like this:

```
aa   aaa      a
  bb    cc  bb
          dd
```

(Sync subtrees would be stacked right below each interval).

---

To get an icicle graph, you'd squish everything to the same baseline (like in Tetris, but going up the page):

```aabbaaaccddbba```

(Sync subtrees would all start at row 1, or row 0 when they ever occur without an async parent).

---

This could be implemented by: Non-async tasks go in row `parent_row+1`; *all async tasks go in row 0*.

(This works because an async task never has an active parent, which can be derived from two properties:
- There is never more than one async task active at a time (by nature of async), so an async task can't have an active async ancestor.
- If an async task has a non-async parent, the parent started it with `spawn`[^1]; in this case, the parent will finish before the executor gets a chance to start the child.)

[^1]: the alternative, `block_on` would be a different story, but we don't use it outside tests and surely never will; it's extremely hazardous for performance.

File: lib/rust/profiler/data/src/lib.rs
Comment (kazcw) [inline lib/rust/profiler/data/src/lib.rs]: `profiler::internal::Timestamp` is a low-level detail data consumers shouldn't need to deal with; we could provide `Mark::into_ms()`.

File: lib/rust/profiler/flame-graph/src/lib.rs
Comment (MichaelMauderer) [reply lib/rust/profiler/flame-graph/src/lib.rs:51]: Good point. I changed the code to align the async blocks in row 0. This seems to work fine for my test.

File: lib/rust/profiler/data/src/lib.rs
Comment (MichaelMauderer) [reply lib/rust/profiler/data/src/lib.rs]: Makes sense. Changed to expose/use `Mark::into_ms().`

**Review by kazcw — CHANGES_REQUESTED** [review]
Now the async intervals will be drawn in the right place, but I think their non-async children will be floating below--they are drawn at their hierarchy level, which is no longer the same as `parent_row+1`.

The test case could use some `async fn` coverage.

**Comment by MichaelMauderer** [comment]
> Now the async intervals will be drawn in the right place, but I think their non-async children will be floating below--they are drawn at their hierarchy level, which is no longer the same as `parent_row+1`.
> 
> The test case could use some `async fn` coverage.

Good point! I’ve updates this to behave correctly now and extended the demo scene to include both calls from sync to async and async to sync.

**Review by kazcw — APPROVED** [review]
Looks good. I can't wait to hook this up to some real data.

**Comment by MichaelMauderer** [comment]
@farmaazon @mwu-tow @4e6 This still requires a code-owner review for the lock file.

**Review by mwu-tow — APPROVED** [review]

**Pull request merged**
